### ISSUE TITLE
cd function can cause what appears to be a memory leak

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#2836]
My [pull-request](https://github.com/fish-shell/fish-shell/pull/2819) to improve support for the C locale modified the **wgetcwd()** function to return a wcstring object rather than an array of wchar_t's, among other things. I was worried that my change might introduce a memory leak. There is a memory leak but the rate at which memory is leaked appears to be unchanged from the current git head. That is, fish currently leaks so much memory I can't determine if my change introduces a new leak.

### Reproduction Steps:

Create a file with the following content (change the shebang line appropriately for you system) and execute the file. In another window monitor the size of the fish process; e.g., `ps -o rss,vsz -fp 16295`.

```
#!/usr/local/bin/fish
while true
    pushd /tmp/Äpfel
    pwd
    echo $PWD
    popd
    pwd
    echo $PWD
    echo pid %self
    sleep 0.1
end
```

### Expected behavior:

No change in the virtual or resident-set size of the process after a few seconds.

### Observed behavior:

The fish process grows at a rate of approximately 9 KB per second on my system.

### Additional information:

Fish version: git head as of the time this issue was opened.

Operating system: OS X 10.11 (El Capitan) with fish built via "make install" which used the Homebrew install GNU build tools (e.g., g++) to compile the program.

**Comment by ridiculousfish** [comment]
Oh good find. I'm able to record allocations in Instruments, and it's very weird - there's very little pattern and no persistent allocations. That is, it looks like all of the memory we allocate gets freed. That suggests it may just be really terrible fragmentation. (Or I could also be misinterpreting or misusing this tool).

**Comment by ridiculousfish** [comment]
`vmmap` confirms this - our malloc zone is very large and almost entirely empty. The issue is that the `$dirprev` global variable is growing very large. You can see that by adding an `echo $dirprev` at the bottom of the loop.

`$dirprev` is set by the cd function. Probably the simplest thing to do is to limit the length of `$dirprev` to 256 entries or something.

**Comment by krader1961** [comment]
Good analysis, @ridiculousfish. I definitely consider it a bug that `$dirprev` grows without bound. Given how the var is used it seems to me a low limit is reasonable. I've never found the nextd/prevd and pushd/popd mechanisms easy enough to reason about to use. Instead, when running ksh and zsh I have a `mcd` function; short for "menu cd". It does the equivalent of displaying the values of the fish `$dirprev` and `$dirnext` lists and asking the user to choose an entry by number. I wrapped the `cd` command to maintain a simple stack of the 25 most recently visited directories. It seems to me a limit of that magnitude (i.e., somewhere between 10 and 100) is appropriate.

**Comment by krader1961** [comment]
Fixed via merge of 9d2b53450ac038bcef7c62191770e9b85b6017f9
